pkg/controller/master/node/promote_controller.go

```diff
@@ -246,53 +247,67 @@ func (h *PromoteHandler) setPromoteResult(job *batchv1.Job, node *corev1.Node, s
 func selectPromoteNode(nodeList []*corev1.Node) *corev1.Node {
 	var (
-		managementNumber     int
-		managementRoleNumber int
-		canPromoteNumber     int
-		promoteNode          *corev1.Node
+		managementNumber               int
+		promoteNode                    *corev1.Node
+		managementNodeTopologyZoneList []string
 	)
 
-	promoteNode = nil
 	for _, node := range nodeList {
-		if isManagementRole(node) {
-			managementNumber++
-			managementRoleNumber++
-		} else if hasPromoteStatus(node) {
+		if isPromoteStatusIn(node, PromoteStatusRunning, PromoteStatusFailed, PromoteStatusUnknown) {
+			// wait until the node promotion is completed or the failed or unknown status is cleared
+			return nil
+		} else if !isManagementRole(node) && isPromoteStatusIn(node, PromoteStatusComplete) {
+			// worker promotion is complete but node is not yet labeled as a management node
+			return nil
+		} else if isManagementRole(node) {
 			managementNumber++
-		} else if isHarvesterNode(node) && isHealthyNode(node) {
-			canPromoteNumber++
-			if promoteNode == nil || node.CreationTimestamp.Before(&promoteNode.CreationTimestamp) {
-				promoteNode = node
+
+			if topologyZone := node.Labels[corev1.LabelTopologyZone]; topologyZone != "" {
+				managementNodeTopologyZoneList = append(managementNodeTopologyZoneList, topologyZone)
 			}
 		}
-	}
 
-	// waiting for the other node completed
-	if managementRoleNumber != managementNumber {
-		return nil
+		// return if there are already enough management nodes
+		if managementNumber == defaultSpecManagementNumber {
+			return nil
+		}
 	}
 
-	// there is no need to promote if the spec number has been reached
-	specManagementNumber := getSpecManagementNumber(len(nodeList))
-	promoteNodeNumber := specManagementNumber - managementNumber
-	if promoteNodeNumber <= 0 {
+	// return if the management node count is equal to the total amount of nodes (there are no more nodes left to promote)
+	nodeCount := len(nodeList)
+	if managementNumber == nodeCount {
 		return nil
 	}
 
-	// make sure have enough nodes can be promote
-	if canPromoteNumber < promoteNodeNumber {
-		return nil
-	}
+	promoteNode = nil
+	for _, node := range nodeList {
+		if !isHealthyNode(node) {
+			// decrease the nodeCount if a node is not healthy
+			nodeCount--
+		}
 
-	return promoteNode
-}
+		// return if the nodeCount is below the defaultSpecManagementNumber
+		if nodeCount < defaultSpecManagementNumber {
+			return nil
+		}
 
-// getSpecMasterNumber get spec management number by all node number
-func getSpecManagementNumber(nodeNumber int) int {
-	if nodeNumber < 3 {
-		return 1
+		if !isManagementRole(node) && isHarvesterNode(node) && isHealthyNode(node) {
+			if len(managementNodeTopologyZoneList) > 0 {
+				// only promote the node if the label is set to a new zone
+				if topologyZone := node.Labels[corev1.LabelTopologyZone]; topologyZone != "" &&
+					!slice.ContainsString(managementNodeTopologyZoneList, topologyZone) {
```

Comment:
strictly speaking, here it still can not return, need to get the smallest timestamp NODE in same new zone

Reply:
Yes, looks like this needs to be changed.

Reply:
This should work?

Before:
```
			if topologyZone := node.Labels[corev1.LabelTopologyZone]; topologyZone != "" &&
				!slice.ContainsString(managementTopologyZoneList, topologyZone) {
				return node
			}
```
After:
```
			if topologyZone := node.Labels[corev1.LabelTopologyZone]; topologyZone != "" &&
				slice.ContainsString(managementTopologyZoneList, topologyZone) {
				continue
			}
```

Reply:
@janeczku 
in selectPromoteNode:
```go
		if !isManagementRole(node) && isHarvesterNode(node) && isHealthyNode(node) {
			topologyZone := node.Labels[corev1.LabelTopologyZone]
			if len(managementNodeTopologyZoneList) > 0 && topologyZone != "" && !slice.ContainsString(managementNodeTopologyZoneList, topologyZone) || len(managementNodeTopologyZoneList) == 0 {
				if promoteNode == nil || node.CreationTimestamp.Before(&promoteNode.CreationTimestamp) {
					promoteNode = node
				}
			}
		}
```


add unit test:
```go
	w4z3 = NewDefaultNodeBuilder().Name("w-4-z3").Zone("zone3").Harvester().Worker()
```
```go
		{
			name: "two management in zone1 and zone2 and two worker in zone3",
			args: args{
				nodeList: []*corev1.Node{m1z1, m2z2, w4z3, w3z3},
			},
			want: w3z3,
		},
```

Reply:
added commit https://github.com/harvester/harvester/pull/2339/commits/5b39f8d31fdba5df986428041b8e9edefc7e2855 to this PR

Reply:
LGTM